tutorials/v7/ntuple/ntpl013_staged.C

```diff
@@ -0,0 +1,205 @@
+{
+   // static variables that are shared between threads; this is done for simplicity in this tutorial, use proper data
+   // structures in real code!
+   static std::mutex g_Mutex;
+   static NTupleSize_t g_WrittenEntries = 0;
+
+   using generator = std::mt19937;
+   generator gen;
+
+   // Create a fill context and turn on staged cluster committing.
+   auto fillContext = writer->CreateFillContext();
+   fillContext->EnableStagedClusterCommitting();
+   auto entry = fillContext->CreateEntry();
+
+   auto eventId = entry->GetPtr<std::uint32_t>("eventId");
+   auto eventIdStart = id * kNEventsPerThread;
+   int startOfBlock = 0;
+   auto rndm = entry->GetPtr<float>("rndm");
+
+   for (int i = 0; i < kNEventsPerThread; i++) {
+      // Prepare the entry with an id and a random number.
+      *eventId = eventIdStart + i;
+      auto d = static_cast<double>(gen()) / generator::max();
+      *rndm = static_cast<float>(d);
+
+      // Fill might auto-flush a cluster, which will be staged.
+      fillContext->Fill(*entry);
+
+      if ((i + 1) % kNEventsPerBlock == 0) {
+         // Decide to flush this cluster and logically append all staged clusters to the ntuple.
+         fillContext->FlushCluster();
+         {
+            std::lock_guard g(g_Mutex);
+            fillContext->CommitStagedClusters();
+            auto firstEvent = eventIdStart + startOfBlock;
+            auto lastEvent = eventIdStart + i;
+            std::cout << "Thread #" << id << " wrote events #" << firstEvent << " - #" << lastEvent << " as entries #"
+                      << g_WrittenEntries << " - #" << (g_WrittenEntries + kNEventsPerBlock - 1) << std::endl;
+            g_WrittenEntries += kNEventsPerBlock;
+            startOfBlock += kNEventsPerBlock;
+         }
+      }
+   }
+
+   // Flush the remaining data and commit staged clusters.
+   fillContext->FlushCluster();
+   {
+      std::lock_guard g(g_Mutex);
+      fillContext->CommitStagedClusters();
+      auto firstEvent = eventIdStart + startOfBlock;
+      auto lastEvent = eventIdStart + kNEventsPerThread - 1;
+      auto numEvents = kNEventsPerThread - startOfBlock;
+      std::cout << "Thread #" << id << " wrote events #" << firstEvent << " - #" << lastEvent << " as entries #"
+                << g_WrittenEntries << " - #" << (g_WrittenEntries + numEvents - 1) << std::endl;
+      g_WrittenEntries += numEvents;
+   }
+}
+
+// Generate kNEvents with multiple threads in kNTupleFileName, and sequence them in blocks of kNEventsPerBlock entries
+void WriteInBlocks()
+{
+   std::cout << "\n === ... with sequencing in blocks of " << kNEventsPerBlock << " events ===" << std::endl;
+
+   // Create the data model
+   auto model = RNTupleModel::CreateBare();
+   model->MakeField<std::uint32_t>("eventId");
+   model->MakeField<float>("rndm");
+
+   // Create RNTupleWriteOptions to make the writing commit multiple clusters.
+   RNTupleWriteOptions options;
+   options.SetApproxZippedClusterSize(32 * 1024);
```

Comment:
The idea is to show that it's possible to mix implicitly and explicitly flushed clusters, so I'd like to keep it. With a single cluster of exactly 10'000 entries, we wouldn't need staged cluster committing for the purpose shown in this tutorial, strictly speaking, because a single cluster (= block) would atomically order with respect to other threads. The problem that we're trying to solve here is sequencing of multiple clusters that collectively make up a block. I added some comments to explain this.